webhooks/unifiedstoragequota/validation/unifiedstoragequota_validator.go

```diff
@@ -0,0 +1,332 @@
+		WebhookContext: h.WebhookContext,
+		Op:             req.Operation,
+		Obj:            obj,
+		OldObj:         oldObj,
+		UserInfo:       req.UserInfo,
+		Logger:         h.WebhookContext.Logger.WithName(obj.GetNamespace()).WithName(obj.GetName()),
+	}
+
+	return handleRequest(webhookRequestContext)
+}
+
+func (h *RequestedCapacityHandler) HandleCreate(ctx *pkgctx.WebhookRequestContext) CapacityResponse {
+	vm := &vmopv1.VirtualMachine{}
+	if err := h.Converter.FromUnstructured(ctx.Obj.UnstructuredContent(), vm); err != nil {
+		return CapacityResponse{Response: webhook.Errored(http.StatusBadRequest, err)}
+	}
+
+	scName := vm.Spec.StorageClass
+
+	sc := &storagev1.StorageClass{}
+	if err := h.Client.Get(ctx, client.ObjectKey{Name: scName}, sc); err != nil {
+		if apierrors.IsNotFound(err) {
+			return CapacityResponse{Response: webhook.Errored(http.StatusNotFound, err)}
+		}
+		return CapacityResponse{Response: webhook.Errored(http.StatusInternalServerError, err)}
+	}
+
+	var capacity *resource.Quantity
+
+	if vm.Status.Storage == nil || !conditions.IsTrue(vm, vmopv1.VirtualMachineConditionCreated) {
+		if vm.Spec.Image == nil {
+			return CapacityResponse{Response: webhook.Errored(http.StatusBadRequest, errors.New("vm.Spec.Image is required"))}
+		}
+		vmiName := vm.Spec.Image.Name
+		var imageStatus vmopv1.VirtualMachineImageStatus
+
+		switch vm.Spec.Image.Kind {
+		case "VirtualMachineImage":
+			vmi := &vmopv1.VirtualMachineImage{}
+			if err := h.Client.Get(ctx, client.ObjectKey{Namespace: vm.Namespace, Name: vmiName}, vmi); err != nil {
+				if apierrors.IsNotFound(err) {
+					return CapacityResponse{Response: webhook.Errored(http.StatusNotFound, err)}
+				}
+				return CapacityResponse{Response: webhook.Errored(http.StatusInternalServerError, err)}
+			}
+			imageStatus = vmi.Status
+		case "ClusterVirtualMachineImage":
+			cvmi := &vmopv1.ClusterVirtualMachineImage{}
+			if err := h.Client.Get(ctx, client.ObjectKey{Name: vmiName}, cvmi); err != nil {
+				if apierrors.IsNotFound(err) {
+					return CapacityResponse{Response: webhook.Errored(http.StatusNotFound, err)}
+				}
+				return CapacityResponse{Response: webhook.Errored(http.StatusInternalServerError, err)}
+			}
+			imageStatus = cvmi.Status
+		}
+
+		if len(imageStatus.Disks) < 1 || imageStatus.Disks[0].Capacity == nil {
+			return CapacityResponse{Response: webhook.Errored(http.StatusNotFound, errors.New("boot disk not found in image status"))}
+		}
+		capacity = imageStatus.Disks[0].Capacity
+	} else {
+		capacity = resource.NewQuantity(0, resource.BinarySI)
+
+		if vm.Status.Storage.Unshared != nil {
+			capacity.Add(*vm.Status.Storage.Unshared)
+		}
+		// Subtract the PVC usage/limit from the total used value as PVCs
+		// are counted separately.
+		for _, volume := range vm.Status.Volumes {
+			if volume.Type == vmopv1.VirtualMachineStorageDiskTypeManaged {
+				if volume.Used != nil {
+					capacity.Sub(*volume.Used)
+				}
+			}
+		}
+	}
```

Comment:
Would this ever be the case for VM creation requests here? IOW, isn't the status field always empty during resource creation and only populated by controllers afterward?

Reply:
@akutz can you comment on this?

Reply:
reworked the logic here